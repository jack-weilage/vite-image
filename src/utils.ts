import type { ImageConfig, Transformer, PluginConfig } from '../types'
import type { Sharp, Metadata } from 'sharp'

import { type BinaryLike, createHash } from 'crypto'
import { basename, extname } from 'path'
import { CONFIG_SCHEMA } from './constants'

/** Limit a number to between min and max. */
export const minmax = (num: number, min: number, max: number) => Math.max(Math.min(num, max), min)
/** 
 * Create a SHA1 hash from a string.
 * 
 * SHA256 is more secure, but speed matters more here.
*/
export const create_hash = (str: BinaryLike): string => createHash('sha1').update(str).digest('hex')
/** Extract the name of a file, without its extension. */
export const filename = (path: string) => basename(path, extname(path))
/** De-duplicate an array. */
export const dedupe = <T>(arr: T[]) => [ ...new Set(arr) ]

/** Create a `Partial` of any object. */
export function copy_only_keys<T>(obj: T, keys: (keyof T)[]): Partial<T> {
    const partial = {} as Partial<T>
    for (const key of keys)
        partial[key] = obj[key]

    return partial
}

/** Validate and combine a user config and default config. */
export function parse_config(user_config: Partial<PluginConfig>, default_config: PluginConfig): PluginConfig
{
    const config = Object.assign({ ...default_config }, user_config)

    const errors = CONFIG_SCHEMA.validate(config)
    if (errors.length !== 0)
        throw new AggregateError(errors)
    
    return config
}

/** Coerces values to string | number | boolean. */
const format_value = (val: string) => {
    if (val === '' || val === 'true')
        return true

    if (val === 'false')
        return false
    
    if (!Number.isNaN(+val))
        return +val

    return val
}

//TODO: Handle `!blur` as `blur=false`
/** Create unique configs from arrays of values. */
export function create_configs(params: URLSearchParams, deliminator: string): ImageConfig[]
{
    const aggregated = {} as Record<string, (string | number | boolean)[]>
    for (const key of dedupe([ ...params.keys() ]))
    {
        // Get every occurrence of the key, then split by the deliminator.
        const values = params.getAll(key)
            .flatMap(value => value.split(deliminator))
            .map(format_value)

        aggregated[key] = dedupe(values)
    }
    
    const keys = Object.keys(aggregated)
    const values = Object.values(aggregated)

    // If there is only one key, the code below won't work.
    if (keys.length === 1)
        return values[0].map(value => ({ [keys[0]]: value }))
    
    // Create an array of unique configs.
    // Each value will have the index of the corresponding keys.
    const groups = values
        //@ts-expect-error: This reducer will take in (string | number | boolean)[] and return (string | number | boolean)[][]
        .reduce((acc, cur) => acc.flatMap(a => cur.map(b => [ a, b ].flat()))) as any as (string | number | boolean)[][]
    
    const configs = [] as ImageConfig[]
    for (const group of groups) {
        const config = {} as Record<string, string | number | boolean>

        // For every value, assign the corresponding key.
        for (let i = 0; i < keys.length; i++)
            config[keys[i]] = group[i]
        
        configs.push(config)
    }

    return configs
}

/** Apply all transformers to an image. */
export function apply_transformers(image: Sharp, metadata: Metadata, config: ImageConfig, transformers: Transformer[])
{
    const applied_transformers = []

    for (const { matcher, transform, name } of transformers)
    {
        if (!matcher(config))
            continue
    
        applied_transformers.push(name)
        image = transform(image, config, metadata)
    }

    return { image, applied_transformers }
}