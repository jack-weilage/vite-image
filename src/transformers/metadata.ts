import type { Transformer } from '../../types'

import { is_boolean } from '../utils'

export default {
	name: 'metadata',
	matcher: (config) => is_boolean(config['metadata']),
	transform: (img, config) => (config['metadata'] ? img.withMetadata() : img),
} satisfies Transformer<'metadata'>
