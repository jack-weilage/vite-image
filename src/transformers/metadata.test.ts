import { it } from 'vitest'

import { base_hash, test_transformer } from '../../tests/utils'
import metadata from './metadata'

it('applies metadata when it should', ({ expect }) => Promise.all([
    // metadata=true
    expect(test_transformer({ metadata: true }, metadata))
        .resolves.toMatchInlineSnapshot('"0b95256618bd832e887e1169cacde6e29bee2ac2"'),
    // metadata=false
    expect(test_transformer({ metadata: false }, metadata))
        .resolves.toBe(base_hash)
]))
it('doesn\'t apply metadata when it shouldn\'t', ({ expect }) => Promise.all([
    // metadata=foo
    expect(test_transformer({ metadata: 'foo' }, metadata))
        .resolves.toBe(base_hash),
    // metadata=1
    expect(test_transformer({ metadata: 1 }, metadata))
        .resolves.toBe(base_hash),
    // metadata=0
    expect(test_transformer({ metadata: 0 }, metadata))
        .resolves.toBe(base_hash)
]))