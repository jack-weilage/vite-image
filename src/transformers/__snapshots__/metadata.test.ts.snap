// Vitest Snapshot v1

exports[`applies the transform metadata=false 1`] = `"64db99769cd8b1dfe79a5e5d2ab0359623103ddd"`;

exports[`applies the transform metadata=true 1`] = `"0b95256618bd832e887e1169cacde6e29bee2ac2"`;
